server/test/integration/dbTest/beginTransactionTest.js

```diff
@@ -0,0 +1,112 @@
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sandbox.stub() }
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).rejects(new Error('deadlock detected')).onCall(1).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should log the deadlock error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the runBeginTransactionWithRetries query: Error: deadlock detected`)
+    })
+
+    it('should log the retry', () => {
+      expect(helpers.log).to.be.calledWith(`Retrying runBeginTransactionWithRetries.`)
+    })
+
+    it('should lock the clients, sessions, and locations tables to reflect the retry', () => {
+      expect(clientStub.query).to.be.calledWith(`LOCK TABLE clients, sessions, locations`)
+    })
+
+    it('should return a valid pgClient', () => {
+      expect(pgClient).to.be.equal(clientStub)
+    })
+  })
+
+  describe('when beginTransaction is successful the first time it is called', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sinon.stub() }
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should not log any errors', () => {
+      expect(helpers.logError).to.not.be.called
+    })
+
+    it('should not log a retry', () => {
+      expect(helpers.log).to.not.be.calledWith(`Retrying runBeginTransactionWithRetries.`)
+    })
+
+    it('should lock the clients, sessions, and locations tables', () => {
+      expect(clientStub.query).to.be.calledWith(`LOCK TABLE clients, sessions, locations`)
+    })
+
+    it('should return a valid pgClient', () => {
+      expect(pgClient).to.be.equal(clientStub)
+    })
+  })
+
+  describe('when beginTransaction fails twice, deadlocking on the first attempt and throwing an error with a valid pgClient on the second attempt', () => {
+    let clientStub
+    beforeEach(async () => {
+      clientStub = { query: sandbox.stub() }
+      clientStub.query.withArgs('BEGIN').rejects(new Error('some error'))
+      sandbox.spy(db, 'beginTransaction')
+      poolConnectStub.onCall(0).rejects(new Error('deadlock detected')).onCall(1).resolves(clientStub)
+      pgClient = await db.beginTransaction()
+    })
+
+    it('should log the deadlock error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the runBeginTransactionWithRetries query: Error: deadlock detected`)
+    })
+
+    it('should log the retry', () => {
+      expect(helpers.log).to.be.calledWith(`Retrying runBeginTransactionWithRetries.`)
+    })
+
+    it('should log the second error', () => {
+      expect(helpers.logError).to.be.calledWith(`Error running the runBeginTransactionWithRetries query: Error: some error`)
+    })
+
+    it('should return null', () => {
+      expect(pgClient).to.be.null
+    })
```

Comment:
In this case, we're also going to try to rollback the transaction right? You could add a test for that case too.

Reply:
Oh this is a good point! My thinking was that because I hadn't made any changes to the rollback code (besides moving it up a few lines) that I wouldn't need to test it. But maybe I should just to be safe.

Reply:
Added! This was a good flag because (as per our discussion in Slack) we realized `this.rollbackTransaction(pgClient)` should actually be `rollbackTransaction(pgClient)` since `this` isn't referring to a class, just the module `db.js`. 

Thanks @trisapeace!

Reply:
Thanks you! I'm glad we found that. It might also have led to some of the weirdness around deadlocks.

What might be fun now is to test on your local machine. You can turn on your DB Logging and make it so that an error is always thrown for the first query and then look in the logs to make sure that it does the rollback.

Reply:
As per our discussion, the using the debugger (when the rollbackStub wasn't working at first) confirmed that `rollbackTransaction()` was in fact being called hitting all the code within the function. So I don't think I'll need to do this as well. Thanks for all your support Theresa!